### ISSUE TITLE
Volume rendering tests broken

### ISSUE THREAD

**Issue by medihack** [x3dom/x3dom#735]
I am trying to get some of the volume rendering stuff in x3dom to work. Unfortunately, many of those features are not well documented. So I thought I could look at some of the [tests (functional/volren)](https://github.com/x3dom/x3dom/tree/master/test/functional/volren), but it seems that those are broken as they link wrong relative locations of the Javascript files. But even after fixing this, I couldn't get many things to work (like ShadedVolumeStyle, [see here](http://stackoverflow.com/questions/43412536/volume-rendering-with-color-and-light-in-x3dom)). So I wonder a bit, what is the status of volume rendering in x3dom? Is it still maintained?

**Comment by andreasplesch** [comment]
Vicomtech contributed the volume rendering component in x3dom. Here is an older demo page: 

http://demos.vicomtech.org/

You could use a transfer function to map from grey to color.

The ShadedVolumeStyle scene works when you paste the x3d below here

http://andreasplesch.github.io/Library/Viewer/x3dom.html

```
<X3D xmlns='http://www.web3d.org/specifications/x3d-namespace' showStat='true' showLog='true' width='500px' height='500px'>
  <Scene>
    <Background skyColor='0 0 0'></Background>
    <Viewpoint id='front' position='0 0 300'></Viewpoint>
    
    <!--<PointLight id='point' on='TRUE' intensity='0.9000' ambientIntensity='0.0000' color='0.0 0.6 0.0' location='0 0 300' attenuation='0 0 0' radius='5.0000'></PointLight>-->
    
    <VolumeData id='brainVolume' dimensions='192 228 192'>
      <ImageTextureAtlas containerField='voxels' url="https://i.imgur.com/3Ydy8cQ.jpg" numberOfSlices='193' slicesOverX='14' slicesOverY='14'></ImageTextureAtlas>
      <ShadedVolumeStyle lighting='TRUE' shadows='TRUE'>
        <Material diffuseColor='1.0 0 0' specularColor='1 0 0' ambientIntensity='0.4'></Material>
      </ShadedVolumeStyle> 
    </VolumeData>
  </Scene>
</X3D>
```
Maybe codepen has some issues.

https://doc.x3dom.org/author/VolumeRendering/ShadedVolumeStyle.html says
that lighting and shadows is not yet implemented but the code definitely uses the lights in the shader:

https://github.com/x3dom/x3dom/blob/master/src/nodes/VolumeRendering/ShadedVolumeStyle.js#L267

But enabling a light gives me also a black screen.

**Comment by andreasplesch** [comment]
jsfiddle seems to work better
https://jsfiddle.net/jyxr05u8/4/

**Comment by aarbelaiz** [comment]
There is only very basic support for lightning with the __ShadedVolumeStyle__ which uses phong based illumination (no shadows), finding the correct parameter values can be tricky. All the other volume nodes should work as expected.

The (functional/volren) are legacy files, they are not x3dom compatible, for volren examples look-up the official [examples](https://www.x3dom.org/examples/) page, the regression-suite tests and the Vicomtech  [demo](http://volumerc.org/demos.html) page.

The best way to colorize a volume is to define a _transfer function_ with the __OpacityMapVolumeStyle__ node. A _transfer function_ is an image of 256 pixels wide and >= 1 pixels height that maps each data value to a given color and opacity. Check the [volren example](https://examples.x3dom.org/volren/volrenOpacityTestTF_aorta.xhtml) source.

I am working in a web based x3dom compatible _transfer function_ editor right now, this would help with the process of adding color to the volume. I may contribute soon with an update of the main volren example as soon as I have a public prototype.

**Comment by medihack** [comment]
First of all, thanks a lot for your help.
In the above jsfiddle the `<Material>` tag is commented out. And if I uncomment it I would have expected that the color of the brain does somehow change. But it doesn't. So I don't see what ShadedVolumeStyle does at all when it does not even respect the provided material.
Maybe it would be clearer if those legacy files would be moved to a "legacy" folder with a note that these might not work. And also the documentation could here be a bit clearer (I wasn't even initially aware of what NYI means ;-)).
So I guess the conclusion is (at least in my case) to use `OpacityMapVolumeStyle` with a _transfer function_ to present several volumes with different colors.

**Comment by andreasplesch** [comment]
Yeah, I commented the material node as a test because I had the same thought. You are right, the material color should be used. You could take a look at the shader code and try to see what is going on with lights and color. It may only need a little tweaking.

**Comment by aarbelaiz** [comment]
@medihack Maybe this can help you to create a transfer function, I just pushed the first version, so do not expect it to be faultless [tf-editor](https://volumerc.github.io/tf-editor/components/tf-editor/) (look for the demo section).

You can look-up [here](https://github.com/VolumeRC/tf-editor-examples/tree/master/web-page) for a simple web page example with x3dom.
